move tailsitter-specific code to new AttitudeControl_TS subclass

kd0aij opened the pull request:
replaces https://github.com/ArduPilot/ardupilot/pull/15195

Adds an overloaded version of relax_attitude_controllers with one boolean argument which simply calls the original no-arg version in the base class.  The new subclass inherits from AC_AttitudeControl_Multi and overrides the one-argument version of  relax_attitude_controllers to implement special behavior for tilt-vectored tailsitters.

The implementation of the existing tailsitter specific method input_euler_rate_yaw_euler_angle_pitch_bf_roll is also moved from the base class to the subclass.

This is a fix for the problem referenced in ArduPilot/ardupilot_wiki#2990
It allows the motors to tilt up in all vtol modes (not just qstabilize) prior to raising the throttle, and prevents prop strikes on arming.

kd0aij commented:
@tridge This is the alternative approach to tailsitter attitude control  specialization preferred by both Randy and Leonard

kd0aij commented:
this fix needs to be rebased on https://github.com/ArduPilot/ardupilot/pull/15301 once that goes in

kd0aij commented:
@Hwurzburg I just retested this in sitl with the SkyCat2. Seems to behave the same as the original branch you flight tested.

kd0aij commented:
@IamPete1 It looks like this branch always had the fix to hold_stabilize, unless I"m confused.

kd0aij commented:
rebased
build/Pixracer/bin/arduplane.apj:    "image_size": 1429312,
master: "image_size": 1428688, delta 624 bytes

kd0aij commented on ArduPlane/quadplane.cpp:
```diff
@@ -708,7 +709,11 @@ bool QuadPlane::setup(void)
     }
 
-    attitude_control = new AC_AttitudeControl_Multi(*ahrs_view, aparm, *motors, loop_delta_t);
+    if (motor_class == AP_Motors::MOTOR_FRAME_TAILSITTER) {
```
does this need to include copter tailsitters also?

kd0aij replied:
yes, fixed

kd0aij commented:
@robustini I think this is the version of https://github.com/ArduPilot/ardupilot/pull/15195 which may eventually go into master. I'd appreciate it if you could test to see whether there is any breakage for non-vectored tailsitters, and what issues remain to be solved.
I'll look into negating any ill effects from Q_TRIM_PITCH on takeoff.

robustini commented:
> 
> 
> @robustini I think this is the version of #15195 which may eventually go into master. I'd appreciate it if you could test to see whether there is any breakage for non-vectored tailsitters, and what issues remain to be solved.
> I'll look into negating any ill effects from Q_TRIM_PITCH on takeoff.

No problem found on non-vectored (SITL and Realflight).

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -741,6 +748,11 @@ bool QuadPlane::setup(void)
+    // TODO: update this if servo function assignments change
+    // used by relax_attitude_control() to control special behavior for vectored tailsitters
+    _is_vectored = ((motor_class == AP_Motors::MOTOR_FRAME_TAILSITTER) &&
+           (!is_zero(tailsitter.vectored_hover_gain) && SRV_Channels::function_assigned(SRV_Channel::k_tiltMotorLeft)));
```
since were only doing this one time you might as well also check tiltMotorRight

kd0aij replied:
but you were the one who made me take that check out :)

IamPete1 replied:
Maybe when it was once a loop? I meant check for one or the other. 
```suggestion
           (!is_zero(tailsitter.vectored_hover_gain) && (SRV_Channels::function_assigned(SRV_Channel::k_tiltMotorLeft) || SRV_Channels::function_assigned(SRV_Channel::k_tiltMotorRIght)));
```

kd0aij replied:
I hope I never had that in a loop, but maybe...
What happens if you "commit suggestion". Does it do a weird merge commit? I pushed the "resolve inconsistencies" button recently, and didn't like the result at all.

IamPete1 replied:
You just get a extra commit on top, the message is abit random usually so I tend to then pull it down squash and push back up. 

Maybe it was on some slower check or something, to be honest I have no idea, maybe I'm just fickle ;)

kd0aij replied:
Thanks, I'll give that a try then.

rmackay9 commented on libraries/AC_AttitudeControl/AC_AttitudeControl_TS.h:
```diff
@@ -0,0 +1,25 @@
+public:
+	AC_AttitudeControl_TS(AP_AHRS_View &ahrs, const AP_Vehicle::MultiCopter &aparm, AP_MotorsMulticopter& motors, float dt);
+
+	// empty destructor to suppress compiler warning
```
little white space issue

rmackay9 commented on libraries/AC_AttitudeControl/AC_AttitudeControl_TS.h:
```diff
@@ -0,0 +1,25 @@
+    virtual void relax_attitude_controllers(bool exclude_pitch) override;
+    virtual void input_euler_rate_yaw_euler_angle_pitch_bf_roll(bool plane_controls, float body_roll_cd, float euler_pitch_cd, float euler_yaw_rate_cds) override;
+    void set_is_vectored(bool v) { _is_vectored = v; }
+    bool get_is_vectored(void) { return _is_vectored; }
```
get_is_vectored() could be const

tridge commented on libraries/AC_AttitudeControl/AC_AttitudeControl_TS.h:
```diff
@@ -0,0 +1,25 @@
+    // Relax only the roll and yaw rate controllers if exclude_pitch is true
+    virtual void relax_attitude_controllers(bool exclude_pitch) override;
+    virtual void input_euler_rate_yaw_euler_angle_pitch_bf_roll(bool plane_controls, float body_roll_cd, float euler_pitch_cd, float euler_yaw_rate_cds) override;
+    void set_is_vectored(bool v) { _is_vectored = v; }
```
look to be unused?

rmackay9 commented on ArduPlane/quadplane.h:
```diff
@@ -115,8 +115,11 @@ class QuadPlane
+    bool is_control_surface_tailsitter(void) const;
+
+    // true when flying a tilt-vectored tailsitter
+    bool _is_vectored;
```
as mentioned on the dev call, there are two variables named, "_is_vectored".  one here in the vechicle code and another in the AC_ATtitude controller.

tridge commented on libraries/AC_AttitudeControl/AC_AttitudeControl_TS.cpp:
```diff
@@ -0,0 +1,113 @@
+#include "AC_AttitudeControl_TS.h"
+
```
add a license and comment header

rmackay9 commented:
Just for the record, we decided on the dev call that after the above changes are made this is OK to merge to master.

kd0aij commented:
resolved issues noted in devcall
1) removed unused member/getter/setter
2) added header to new cpp file
3) ran astyle on new files
and squashed
@Hwurzburg Ready for flight testing.

kd0aij commented:
@tridge @rmackay9 the 2 new commits change the base class relax_attitude_controllers() method back to non-virtual
and always instantiate AC_AttitudeControl_TS in quadplane, since the new relax_attitude_controllers(bool) method is always used by quadplane.

kd0aij commented:
@tridge @IamPete1 Henry and I have added another commit to prevent prop strikes when changing to a VTOL mode when armed on the ground. It tests OK in SITL and on one of Henry's TVBS planes.
Not sure why CI is failing now; the failing (copter) tests pass locally, except for replay, and I don't see why replay should be affected by this. Last thing I tried was rebasing on latest master.

kd0aij commented:
rebased again and passing CI
